src/java.base/share/classes/jdk/internal/foreign/Utils.java

```diff
@@ -90,48 +110,57 @@ public static MemorySegment alignUp(MemorySegment ms, long alignment) {
     }
 
     /**
-     * This method returns a <em>raw var handle</em>, that is, a var handle that does not perform any size
-     * or alignment checks. Such checks are added (using adaptation) by {@link LayoutPath#dereferenceHandle()}.
+     * This method returns a var handle that takes a memory segment and offset (MS, long). If it has strides, it also
+     * takes a raw offset (MS, long, long), which the var handle does not perform any size or alignment checks against.
+     * Such checks are added (using adaptation) by {@link LayoutPath#dereferenceHandle()}.
      * <p>
      * We provide two level of caching of the generated var handles. First, the var handle associated
      * with a {@link ValueLayout#varHandle()} call is cached inside a stable field of the value layout implementation.
      * This optimizes common code idioms like {@code JAVA_INT.varHandle().getInt(...)}. A second layer of caching
-     * is then provided by this method: after all, var handles constructed by {@link MemoryLayout#varHandle(PathElement...)}
-     * will be obtained by adapting some raw var handle generated by this method.
+     * is then provided by this method, so different value layouts with same effects can reuse var handle instances.
+     * (The 2nd layer may be redundant in the long run)
      *
-     * @param layout the value layout for which a raw memory segment var handle is to be created.
-     * @return a raw memory segment var handle.
+     * @param enclosing the enclosing context of the value layout
+     * @param layout the value layout for which a raw memory segment var handle is to be created
+     * @param noStride if there is no stride, the VH uses the fixed offset instead of taking an offset
+     * @param offset the offset in case there is no stride
+     * @return a raw memory segment var handle
      */
-    public static VarHandle makeRawSegmentViewVarHandle(ValueLayout layout) {
-        final class VarHandleCache {
-            private static final Map<ValueLayout, VarHandle> HANDLE_MAP = new ConcurrentHashMap<>();
+    public static VarHandle makeRawSegmentViewVarHandle(MemoryLayout enclosing, ValueLayout layout, boolean noStride, long offset) {
+        if (enclosing instanceof ValueLayout direct) {
```

Comment:
For now the caching is ok. Moving forrward, I wonder if we shouldn't add more reuse between field var handles in cases like these:

```
struct Point {
   int x; int y;
}
```

```
struct Tuple {
   int u; int v;
}
```

E.g. if I access the first field in these two structs, then the var handle will be identical? In a way, we use the enclosing layout for two reasons:

* to determine the size `S` of the accessed memory region
* to determine the alignment `A` of the accessed memory region

So, if the static `offset`, accessed `layout`, `S` and `A` are the same for two accesses, we should probably reuse the same var handle instance? E.g. whether we access `x` in a `Point`, or `u` in a `Tuple` makes little difference.

Reply:
I think direct var handles are quite cheap already and don't really need caching. What really worth caching should be the combinators and the VH derived from those combinators.

Reply:
> I think direct var handles are quite cheap already and don't really need caching. What really worth caching should be the combinators and the VH derived from those combinators.

Right -- and one big advantage of this PR, if I'm correct, is that if you have a VH to access a single field, that will be a non-strided VH with a static offset (saved in the VH instance). So no adaptation is required (meaning no spinning of bytecodes) and, also, var handle guards will apply too. This is a very good result!